src/linear_algebra/eigenspace.lean

```diff
@@ -242,5 +248,48 @@ begin
       exact h_lμ_eq_0 μ h_cases } }
 end
 
+/-- The generalized eigenspace for a linear map `f`, a natural number `k`, and a scalar `μ` is the
+    kernel of `(f x - μ • id) ^ k`. -/
+def generalized_eigenspace [comm_ring K] [module K V]
+  (f : End K V) (k : ℕ) (μ : K) : submodule K V :=
+((f - am μ) ^ k).ker
+
+/-- A nonzero element of a generalized eigenspace is a generalized eigenvector. -/
+def has_generalized_eigenvector [comm_ring K] [module K V]
+  (f : End K V) (k : ℕ) (μ : K) (x : V) : Prop :=
+x ≠ 0 ∧ x ∈ generalized_eigenspace f k μ
+
+/-- A scalar `μ` is a generalized eigenvalue for a linear map `f` and a natural number `k` if there
+    are generalized eigenvectors for `f`, `k`, and `μ`. -/
+def has_generalized_eigenvalue [comm_ring K] [module K V]
+  (f : End K V) (k : ℕ) (a : K) : Prop :=
+generalized_eigenspace f k a ≠ ⊥
+
+/-- The natural number of a generalized eigenvalue is never 0. -/
+lemma exp_ne_zero_of_has_generalized_eigenvalue_ne_zero [comm_ring K] [module K V]
+  {f : End K V} {k : ℕ} {μ : K} (h : f.has_generalized_eigenvalue k μ) :
+  k ≠ 0 :=
+begin
+  intro hk,
+  rw hk at h,
+  exact h linear_map.ker_id
+end
+
+/-- A generalized eigenspace for some number `k` is contained in
+    the generalized eigenspace for number larger than `k`. -/
+lemma has_generalized_eigenvalue_zero_beyond [field K] [vector_space K V]
+  {f : End K V} {k : ℕ} {μ : K} {m : ℕ} (hm : k ≤ m) :
+  f.generalized_eigenspace k μ ≤ f.generalized_eigenspace m μ :=
+begin
+  simp only [generalized_eigenspace, ←pow_sub_mul_pow _ hm],
+  apply linear_map.ker_le_ker_comp ((f - am μ) ^ k) ((f - am μ) ^ (m - k))
+end
+
+/-- All eigenvalues are generalized eigenvalues. -/
```

Comment:
The docstring and name suggest `eigenvalue`, but the type is `eigenspace`. Could you also add the corresponding lemmas for `eigenvalue`s?
```suggestion
/-- The eigenspace is a subspace of the generalized eigenspace. -/
```